src/field/FieldEquation.i.hh

```diff
@@ -0,0 +1,73 @@
+//---------------------------------*-CUDA-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file FieldEquation.i.hh
+//---------------------------------------------------------------------------//
+#include "MagField.hh"
+
+#include "base/Constants.hh"
+#include <cmath>
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with a constant magnetic field.
+ */
+CELER_FUNCTION
+FieldEquation::FieldEquation(const MagField& field) : field_(field)
+{
+    // set the default charge and the (Lorentz) coefficent with -eplus
+    set_charge(units::ElementaryCharge{-1.0});
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Set charge for this track and reset the (Lorentz) coefficent
+ */
+
+CELER_FUNCTION void FieldEquation::set_charge(units::ElementaryCharge charge)
+{
+    charge_ = charge;
+    coeffi_ = charge_.value() * constants::c_light * this->scale();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Evaluate the right hand side of the Lorentz equation
+ *
+ * \f[
+    m \frac{d^2 \vec{x}}{d t^2} = (q/c)(\vec{v} \times  \vec{B})
+    s = |v|t
+    \vec{y} = d\vec{x}/ds
+    \frac{d\vec{x}}{ds} = \vec{v}/|v|
+    \frac{d\vec{y}}{ds} = (q/pc)(\vec{y} \times \vec{B})
+   \f]
+ */
+CELER_FUNCTION
+auto FieldEquation::operator()(const OdeState& y) const -> OdeState
+{
+    // Get a magnetic field value at a given position
+    Real3 B = field_({y[0], y[1], y[2]});
+
+    // Evaluate the right-hand-side of the equation
+    OdeState rhs;
+
+    real_type momentum_mag2 = y[3] * y[3] + y[4] * y[4] + y[5] * y[5];
+    CELER_ASSERT(momentum_mag2 > 0.0);
+    real_type momenum_inv = 1.0 / std::sqrt(momentum_mag2);
+
+    rhs[0] = y[3] * momenum_inv;
+    rhs[1] = y[4] * momenum_inv;
+    rhs[2] = y[5] * momenum_inv;
+    rhs[3] = coeffi_ * (y[4] * B[2] - y[5] * B[1]) * momenum_inv;
+    rhs[4] = coeffi_ * (y[5] * B[0] - y[3] * B[2]) * momenum_inv;
+    rhs[5] = coeffi_ * (y[3] * B[1] - y[4] * B[0]) * momenum_inv;
```

Comment:
This is one of those places where the `OdeState` would be improved by making it a struct of extra components:
```c++
struct OdeState
{
    Real3 pos;
    Real3 mom;
};
```
because these lines could be written **much** more readably and safely as
```c++
axpy(momentum_inv, y.momentum, &result.pos);
result.mom.fill(0);
axpy(coeffi_ * momentum_inv, cross_product(y.mom, mag_vec), &result.mom);
```

(P.S. there's a typo in your `momentum_inv` variable)

Reply:
result.mom.fill(0) did not compiled (not modifiable?).  I added a constructor to initialize pos and mon.  The new code looks like
<pre>
    // Get a magnetic field value at a given position                           
    Real3 mag_vec = field_(y.pos);

    real_type momentum_mag2 = dot_product(y.mom, y.mom);
    CELER_ASSERT(momentum_mag2 > 0.0);
    real_type momentum_inv = 1.0 / std::sqrt(momentum_mag2);

    // Evaluate the right-hand-side of the equation                             
    OdeState result;

    axpy(momentum_inv, y.mom, &result.pos);
    axpy(coeffi_ * momentum_inv, cross_product(y.mom, mag_vec), &result.mom);

    return result;

</pre>

Reply:
Oh yeah, sorry about that -- I didn't initially create test for that 😞  but I fixed it in https://github.com/celeritas-project/celeritas/pull/184/commits/7f00ccc1d072a190cdc6475e176601934c8230e0 , so you should be able to merge master into your branch to get `fill` working again.

Reply:
Since OdeState pos needs to be initialized too and there are many places where we need to initialize local OdeState, may it be better to explicitly initialize pos and mom? 
<pre>
struct OdeState
{
    Real3 pos; //!< Particle position                                           
    Real3 mom; //!< Particle momentum                                           

    CELER_FUNCTION
    OdeState() : pos.fill(zero_quantity()), mom.fill(zero_quantity()) {}
};
</pre>

Reply:
Yes, initializing in the constructor is definitely a good call for this situation. I think with C++ initializers this can be written as:
```
struct OdeState
{
    Real3 pos{0, 0, 0}; //!< Particle position                                           
    Real3 mom{0, 0, 0}; //!< Particle momentum
};
```